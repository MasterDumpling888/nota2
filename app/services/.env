GOOGLE_GENAI_API_KEY=AIzaSyDKoSfHEB0oWlEdXNH-BmeFsZQuDoT2Dqs