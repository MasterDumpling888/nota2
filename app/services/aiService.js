import axios from 'axios';

const BACKEND_API_URL = 'http://localhost:3000';

export const summarizeNote = async (content) => {
  try {
    const response = await axios.post(`${BACKEND_API_URL}/summarize`, { content });
    return response.data.summary;
  } catch (error) {
    console.error('Summarization error:', error.response?.data || error.message);
    throw new Error('Failed to connect to the summarization service');
  }
};

export const generateQuestionsAndAnswers = async (content) => {
    try {
     
      const response = await axios.post(`${BACKEND_API_URL}/generate-qa`, { content });
      
      return response.data.qaPairs;
    } catch (error) {
      console.error('Q&A generation error:', error.response?.data || error.message);
      throw new Error('Failed to generate questions and answers');
    }
  };
  
const testAI = async () => {
const content = "The Industrial Revolution, which began in the late 18th century, marked a major turning point in history. Almost every aspect of daily life was influenced in some way. The introduction of new machines, such as the steam engine, revolutionized manufacturing and transportation. Factories emerged, leading to urbanization as people moved to cities in search of work. While the Industrial Revolution brought about significant economic growth and technological advancements, it also had negative consequences, including poor working conditions, child labor, and environmental degradation. Over time, labor laws and reforms were introduced to address these issues, but the impact of this era is still felt today.";
try {
    const summary = await summarizeNote(content);
    console.log('Summary:', summary);

    const qaPairs = await generateQuestionsAndAnswers(content);
    console.log('Questions and Answers:');
    qaPairs.forEach((pair, index) => {
    console.log(`\nQ${index+1}: ${pair.question}`);
    console.log(`A${index+1}: ${pair.answer}`);
    });
} catch (error) {
    console.error('Test failed:', error);
}
};

// Run the test
testAI();