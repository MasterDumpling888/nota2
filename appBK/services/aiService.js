import { GoogleGenerativeAI } from '@google/generative-ai';

// Initialize Gemini
const genAI = new GoogleGenerativeAI('AIzaSyDKoSfHEB0oWlEdXNH-BmeFsZQuDoT2Dqs');
const model = genAI.getGenerativeModel({ model: 'gemini-1.5-flash' });

export const summarizeNote = async (content) => {
  try {
    const prompt = `Summarize the following text in 2-3 sentences:\n\n${content}`;
    const result = await model.generateContent(prompt);
    return result.response.text();
  } catch (error) {
    console.error('Failed to summarize text', error);
    throw new Error('Failed to summarize text');
  }
};

export const generateReviewQuestions = async (content) => {
  try {
    const prompt = `Generate 3 questions and answers based on the following text. Format the response as a JSON array where each object contains a "question" and "answer" field:\n\n${content}`;
    const result = await model.generateContent(prompt);
    const responseText = result.response.text();

    // Parse the response as JSON
    let qaPairs;
    try {
      // Try to extract JSON array from the response
      const jsonMatch = responseText.match(/\[[\s\S]*\]/);
      if (jsonMatch) {
        qaPairs = JSON.parse(jsonMatch[0]);
      } else {
        throw new Error('Invalid JSON format in response');
      }
    } catch (parseError) {
      console.error('Error parsing JSON response:', parseError);
      console.log('Raw response:', responseText);

      // Fallback: Try to format response manually
      // This is a basic fallback in case the model doesn't return proper JSON
      qaPairs = [];
      const sections = responseText.split(/\d+\.\s*Question:/i);
      sections.shift(); // Remove first empty section if any

      for (const section of sections) {
        const questionMatch = section.match(/^(.*?)(?:Answer:)/is);
        const answerMatch = section.match(/Answer:(.*?)(?:\d+\.\s*Question:|$)/is);

        if (questionMatch && answerMatch) {
          qaPairs.push({
            question: questionMatch[1].trim(),
            answer: answerMatch[1].trim()
          });
        }
      }
    }

    return qaPairs;
  } catch (error) {
    console.error('Failed to generate questions and answers', error);
    throw new Error('Failed to generate questions and answers');
  }
};
