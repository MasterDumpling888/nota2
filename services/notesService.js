import { db, auth } from '../firebaseConfig';
import { collection, addDoc, updateDoc, deleteDoc, getDocs, doc, getDoc } from 'firebase/firestore';

const getUserNotesCollection = () => {
  const user = auth.currentUser;
  if (!user) throw new Error('User not authenticated');
  return collection(db, 'users', user.uid, 'notes');
};

const getUserFoldersCollection = () => {
  const user = auth.currentUser;
  if (!user) throw new Error('User not authenticated');
  return collection(db, 'users', user.uid, 'folders');
};

const getUserTagsCollection = () => {
  const user = auth.currentUser;
  if (!user) throw new Error('User not authenticated');
  return collection(db, 'users', user.uid, 'tags');
};

export const addNote = async (note) => {
  try {
    const notesCollection = getUserNotesCollection();
    const docRef = await addDoc(notesCollection, note);
    return docRef.id;
  } catch (error) {
    console.error("Error adding note: ", error);
    throw error;
  }
};

export const updateNote = async (id, updatedNote) => {
  try {
    const notesCollection = getUserNotesCollection();
    const noteDoc = doc(notesCollection, id);
    await updateDoc(noteDoc, updatedNote);
  } catch (error) {
    console.error("Error updating note: ", error);
    throw error;
  }
};

export const deleteNote = async (id) => {
  try {
    const notesCollection = getUserNotesCollection();
    const noteDoc = doc(notesCollection, id);
    await deleteDoc(noteDoc);
  } catch (error) {
    console.error("Error deleting note: ", error);
    throw error;
  }
};

export const fetchNotes = async () => {
  try {
    const notesCollection = getUserNotesCollection();
    const querySnapshot = await getDocs(notesCollection);
    const notesList = querySnapshot.docs.map(doc => ({ id: doc.id, ...doc.data() }));
    return notesList;
  } catch (error) {
    console.error("Error fetching notes: ", error);
    throw error;
  }
};

export const getNote = async (id) => {
  try {
    const notesCollection = getUserNotesCollection();
    const noteDoc = doc(notesCollection, id);
    const noteSnapshot = await getDoc(noteDoc);
    if (noteSnapshot.exists()) {
      return { id: noteSnapshot.id, ...noteSnapshot.data() };
    } else {
      throw new Error('Note not found');
    }
  } catch (error) {
    console.error("Error getting note: ", error);
    throw error;
  }
};

export const addFolder = async (folder) => {
  try {
    const foldersCollection = getUserFoldersCollection();
    const docRef = await addDoc(foldersCollection, folder);
    return docRef.id;
  } catch (error) {
    console.error("Error adding folder: ", error);
    throw error;
  }
};

export const addTag = async (tag) => {
  try {
    const tagsCollection = getUserTagsCollection();
    const docRef = await addDoc(tagsCollection, tag);
    return docRef.id;
  } catch (error) {
    console.error("Error adding tag: ", error);
    throw error;
  }
};

export const fetchFolders = async () => {
  try {
    const foldersCollection = getUserFoldersCollection();
    const querySnapshot = await getDocs(foldersCollection);
    const foldersList = querySnapshot.docs.map(doc => ({ id: doc.id, ...doc.data() }));
    return foldersList;
  } catch (error) {
    console.error("Error fetching folders: ", error);
    throw error;
  }
};

export const fetchTags = async () => {
  try {
    const tagsCollection = getUserTagsCollection();
    const querySnapshot = await getDocs(tagsCollection);
    const tagsList = querySnapshot.docs.map(doc => ({ id: doc.id, ...doc.data() }));
    return tagsList;
  } catch (error) {
    console.error("Error fetching tags: ", error);
    throw error;
  }
};
