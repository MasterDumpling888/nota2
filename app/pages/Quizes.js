import React, { useState, useEffect, useRef, useCallback } from 'react';
import { ScrollView, View, Text, PanResponder, Animated, StyleSheet, Dimensions } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import { useFocusEffect } from '@react-navigation/native'; // Import this!
import PageBox from '../components/PageBox';
import Icon from 'react-native-vector-icons/Ionicons';
import { generateQuestionsAndAnswers } from '../services/aiService';
import { fetchNotes } from '../services/notesService';
import { getFontSize } from '../responsiveFont';

const SCREEN_WIDTH = Dimensions.get('window').width;

function Quizes({ navigation }) {
  const [notes, setNotes] = useState([]);
  const slideAnims = useRef({}); // Store slide animations for each note

  useEffect(() => {
    const loadNotes = async () => {
      try {
        const notesList = await fetchNotes();
        setNotes(notesList);
      } catch (error) {
        console.error('Failed to load notes', error);
      }
    };

    loadNotes();
  }, []);

  // ðŸŸ¢ Reset all sliders when the user returns to this screen
  useFocusEffect(
    useCallback(() => {
      Object.values(slideAnims.current).forEach(anim => {
        anim.setValue(0);
      });
    }, [])
  );

  const handleNoteSelect = async (note) => {
    try {
      const generatedQuiz = await generateQuestionsAndAnswers(note.content);
      navigation.navigate('QuizDetail', { quiz: generatedQuiz });
    } catch (error) {
      console.error('Failed to generate quiz', error);
    }
  };

  return (
    <SafeAreaView style={{ flex: 1, backgroundColor: '#181818' }}>
      <ScrollView showsVerticalScrollIndicator={false}>
        <PageBox title="Select a Note for Quiz" onClose={() => navigation.goBack()}>
          {notes.length === 0 ? (
            <Text type="default">No notes available.</Text>
          ) : (
            notes.map((note, index) => (
              <SlideButton
                key={note.id || index}
                note={note}
                onSelect={handleNoteSelect}
                slideAnims={slideAnims}
              />
            ))
          )}
        </PageBox>
      </ScrollView>
    </SafeAreaView>
  );
}

const SlideButton = ({ note, onSelect, slideAnims }) => {
  if (!slideAnims.current[note.id]) {
    slideAnims.current[note.id] = new Animated.Value(0);
  }

  const slideAnim = slideAnims.current[note.id];

  const panResponder = useRef(
    PanResponder.create({
      onStartShouldSetPanResponder: () => true,
      onPanResponderMove: (e, gestureState) => {
        const maxSlideDistance = SCREEN_WIDTH - 110;
        if (gestureState.dx >= 0 && gestureState.dx <= maxSlideDistance) {
          slideAnim.setValue(gestureState.dx);
        }
      },
      onPanResponderRelease: (e, gestureState) => {
        const maxSlideDistance = SCREEN_WIDTH - 110;
        const halfwayPoint = maxSlideDistance / 2;

        if (gestureState.dx > halfwayPoint) {
          // ðŸŸ¢ Spring to the end & THEN call onSelect
          Animated.spring(slideAnim, {
            toValue: maxSlideDistance,
            useNativeDriver: false,
          }).start(() => {
            onSelect(note);
          });
        } else {
          Animated.spring(slideAnim, {
            toValue: 0,
            useNativeDriver: false,
          }).start();
        }
      },
    })
  ).current;

  return (
    <View style={styles.noteContainer}>
      <Animated.View
        {...panResponder.panHandlers}
        style={[styles.slideButton, { transform: [{ translateX: slideAnim }] }]}
      >
        <Icon name="arrow-forward-outline" size={24} color="black" />
      </Animated.View>
      <View style={styles.slideContent}>
        <Text type="title" style={styles.title}>{note.title}</Text>
        <View style={styles.slideTextContainer}>
          <Icon name="sparkles-outline" size={12} color="#ACACAC" />
          <Text style={styles.slideText}>Slide to Generate Quiz</Text>
        </View>
      </View>
    </View>
  )
};

const styles = StyleSheet.create({
  noteContainer: {
    padding: 10,
    borderWidth: 1,
    borderColor: '#33FD0A',
    borderRadius: 50,
    backgroundColor: "#181818",
    marginBottom: 10,
    flexDirection: 'row',
    alignItems: 'center',
    gap: 10,
  },
  slideButton: {
    padding: 10,
    backgroundColor: '#33FD0A',
    borderRadius: 50,
    alignItems: 'center',
    gap: 10,
    zIndex: 1
  },
  slideContent: {
    flex: 1,
    flexDirection: 'column',
    alignItems: 'flex-start',
    justifyContent: 'center',
  },
  title: {
    flex: 1,
    color: 'white',
    fontSize: getFontSize(16),
    fontWeight: 'bold',
  },
  slideTextContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    gap: 3,
  },
  slideText: {
    color: '#ACACAC',
    fontSize: getFontSize(12),
    fontWeight: 'light',
  }
});

export default Quizes;