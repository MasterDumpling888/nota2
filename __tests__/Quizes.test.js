import React from 'react';
import { render, waitFor } from '@testing-library/react-native';
import Quizes from '../app/pages/Quizes';
import { fetchNotes } from '../app/services/notesService';
import { generateReviewQuestions } from '../app/services/aiService';

// Mock dependencies specific to Quizes test
jest.mock('../app/services/notesService', () => ({
    fetchNotes: jest.fn()
}));
jest.mock('../app/services/aiService', () => ({
    generateReviewQuestions: jest.fn()
}));
jest.mock('@react-navigation/native', () => ({
    useFocusEffect: jest.fn(callback => callback())
}));

// Mock Animated and PanResponder
jest.mock('react-native', () => {
    const reactNative = jest.requireActual('react-native');
    
    reactNative.Animated.Value = function(initialValue) {
        this.value = initialValue;
        this.setValue = jest.fn(newValue => { this.value = newValue; });
        this.spring = jest.fn(() => ({ start: jest.fn(cb => cb && cb()) }));
    };
    
    reactNative.Dimensions.get = jest.fn(() => ({ width: 400, height: 800 }));
    
    reactNative.PanResponder.create = jest.fn(config => ({
        panHandlers: {},
        _config: config
    }));
    
    return reactNative;
});

describe('Quizes Component', () => {

    const mockNotes = [
        { id: '1', title: 'Note 1', content: 'Content 1' },
        { id: '2', title: 'Note 2', content: 'Content 2' }
    ];

    beforeEach(() => {
        jest.clearAllMocks();
        fetchNotes.mockResolvedValue(mockNotes);
        generateReviewQuestions.mockResolvedValue({ questions: [] });
    });

    test('renders correctly with notes', async () => {
        const { getByText } = render(
            <Quizes navigation={global.mockNavigation} />
        );

        // Wait for notes to load
        await waitFor(() => {
            expect(fetchNotes).toHaveBeenCalledTimes(1);
            expect(getByText('Note 1')).toBeTruthy();
            expect(getByText('Note 2')).toBeTruthy();
        });
    });

    test('renders no notes message when notes array is empty', async () => {
        fetchNotes.mockResolvedValue([]);
        
        const { getByText } = render(
            <Quizes navigation={global.mockNavigation} />
        );

        await waitFor(() => {
            expect(getByText('No notes available.')).toBeTruthy();
        });
    });

    // Snapshot test
    test('Quizes snapshot test', () => {
        const { toJSON } = render(
            <Quizes navigation={global.mockNavigation} />
        );
        expect(toJSON()).toMatchSnapshot();
    });
});