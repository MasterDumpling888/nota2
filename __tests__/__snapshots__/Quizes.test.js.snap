// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quizes Component Quizes snapshot test 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "backgroundColor": "#181818",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    showsVerticalScrollIndicator={false}
  >
    <View>
      <PageBox
        onClose={[Function]}
        title="Select a Note for Quiz"
      >
        <Text
          type="default"
        >
          No notes available.
        </Text>
      </PageBox>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
